### ISSUE TITLE
Bucket sampler __len__ gives inconsistent result w.r.t the number of actually sampled batches

### ISSUE THREAD

**Issue by janvainer** [lhotse-speech/lhotse#212]
There seems to be an inconsitency in how length of the lhotse samplers is calculated. Since the dataset returns full batches with a single call to `__getitem__`, the length of the dataset should correspond to the total number of batches that can be sampled without repetition. The dataset length in the DataLoader is accessed through the Sampler, but the length of lhote samplers equals the number of provided cuts, which does not equal the number of batches that will be sampled in one epoch.

Example:
```
batch_size = 16
# sample buckets of size 16
train_sampler = BucketingSampler(cuts, max_frames=float("inf"), max_cuts=batch_size)
train_loader = DataLoader(
       dataset, # created from cuts
       sampler=train_sampler,
       batch_size=None,
)

i = 0
for b in train_loader:
    i += 1

print(i)  # 730
print(len(train_sampler))  # 11611 <- this should probably equal 730
print(len(cuts))  # 11611
```

This is problematic for example when you want to train your model with pytorch lightning. They save checkpoints after several epochs, but with SimpleCutSampler no epoch is ever completed and the checkpoints do not get saved to disk. it is also a bit confusing from the usage standpoint. 
I guess it would be a bit difficult to know ahead of time, how many batches there will be if the SingleCutSampler is used (except for the example shown above). Maybe we could pre-determine (force) how many batches will be sampled with the SImpleCutSampler ahead of time given that we may skip some datapoints in the given epoch. Or we could simulate the sampling process with fixed random seed before each epoch (should be ok for medium sized datasets) and know the number of batches exactly. WYT?

**Comment by pzelasko** [comment]
Thanks for bringing this up! I agree that we should respect these conventions. Your last suggestion seems like a natural extension to the current samplers. I think we can do this by separating the sampling logic from batch iteration, so that we’re able to „simulate” what batches are returned in a full epoch and return an accurate number of batches. The random seed mechanism is actually implemented there already.

**Comment by pzelasko** [comment]
Forgot to ask - would you like to take this one?

**Comment by janvainer** [comment]
Sure, will look into it :)

**Comment by janvainer** [comment]
Added PR with a fix #213

**Comment by pzelasko** [comment]
Solved by #213
